[#41340] Files Tab in work package details

Kharonus opened the pull request:
- https://community.openproject.org/work_packages/41340

TODO:
- [x] find correct places when to fetch the work package attachments
- [x] test solved bug https://community.openproject.org/projects/openproject/work_packages/28306/activity
- [x] test attachment on currently creating workpackages
- [x] figure direct upload

Kharonus committed:
[#41340] migrate attachments to files tab

- https://community.openproject.org/work_packages/41340
- added new files tab to wp details view
- added attachments components to files tab
- removed attachments from wp single view, except for wp creation
- fixed a couple of eslint issues in affected files
- migrated attachments to resource store
- added upload to attachment store
- pull in newly built ckeditor
- removed old, hacky regression test

Kharonus committed:
[#41340] fixed small issues found in PR

oliverguenther commented on frontend/src/app/core/state/attachments/attachments.service.ts:
```diff
@@ -0,0 +1,261 @@
+   */
+  requireCollection(key:string):AttachmentsResourceService {
+    if (this.store.getValue().collections[key]) {
+      return this;
```
I could understand if we return the collection observable here directly for an easy "requireAndStream" alternative call, but why return the service here? You're not using it chained afterwards and it's not a common pattern to do this with services.

Kharonus replied:
The idea was to concat the service call like 

```
let x$ = this.attachmentsResourceService.requireCollection(key).query.select()
```

instead of 

```
if (keyNotLoadedInCollection) {
  this.attachentsResourceService.fetchAttachments(key).subscribe();
}

let x$ = this.attachmentsResourceService.query.select()
```

Sadly, in my only example, the collection can only be marked as required, if I am not on a new resource, so there is still the if.

So, you think this is no common pattern? I'm trying to think of a bit more sofisticated one. Join me please, if you have a great idea.

Kharonus replied:
I removed `return this` and changed return type to void.

I'd say, we need another round to come up with a fine pattern, of how the akita stores and services can implement an equivalent to `requireAndStream`

oliverguenther commented on frontend/src/app/core/state/attachments/attachments.service.ts:
```diff
@@ -0,0 +1,261 @@
+                  ...collections,
+                  [collectionKey]: {
+                    ...collections[collectionKey],
+                    ids: (collections[collectionKey]?.ids || []).filter((id) => id !== attachment.id),
```
It would probably make sense to extract these common CRUD operations on collections into helper methods, as they are quite verbose with akita/immutable objects. Doesn't have to be made in this PR though, just a mental note

Kharonus replied:
There is already one introduced by you I assume, for the insertion (overwrite). Basically in the same way we could add `removeItemFromCollection` and `addToCollection`.

Kharonus replied:
I'd rather take this improvement to another step. There is still the thing about the collection key difference in the services. Most use a collection key created by the URL query params. But attachments are context based collection, that do not have params, but a path that differs.

Here we need a more generic approach to depict both use cases for collections.

oliverguenther replied:
Yes, I agree. Thanks for considering it.

oliverguenther commented on frontend/src/app/core/state/attachments/attachments.service.ts:
```diff
@@ -0,0 +1,261 @@
+                  ...collections,
+                  [collectionKey]: {
+                    ...collections[collectionKey],
+                    ids: (collections[collectionKey]?.ids || []).concat(attachments.map((a) => a.id)),
```
Same here (updating a collection value) wrt. the collection helper method.

Kharonus replied:
see above.

oliverguenther commented on frontend/src/app/features/work-packages/components/wp-single-view/wp-single-view.component.ts:
```diff
@@ -129,12 +129,14 @@ export class WorkPackageSingleViewComponent extends UntilDestroyedMixin implemen
     },
   };
 
+  public get isNewResource():boolean {
```
If possible, avoid getters or function calls from templates and instead, update a field on the component instead whenever the work package updates, as this gets called on every change detection

Kharonus replied:
replaced with public field

Kharonus replied:
Just as a remark: the getters do have a reason at some places. E.g. if the input is put into the component asynchronously, like it is now with the user into the authoring component of the attachment list item, a public field set in ngOnInit is not enough.

Here the field seems to be the right choice, as the work package resource is just passed down from top to bottom through those components.

oliverguenther commented on frontend/src/app/shared/components/attachments/attachment-list/attachment-list-item.component.ts:
```diff
@@ -55,30 +60,47 @@ export class AttachmentListItemComponent {
+    return this.text.removeFile({ fileName: this.attachment.fileName });
+  }
+
+  public $author:Observable<IUser>;
```
The common convention popularized by cyclejs and recommended for RXJS observables is a _trailing_ dollar, e.g., `author$`

Kharonus replied:
ups, my fault. moved `$` around

oliverguenther reviewed (changes requested):
Looking very good overall. I have some minor code remarks, good to merge otherwise 👍

Kharonus committed:
[#41340] renaming and small fixes

- replaced getter with public field
- renamed observable to convention

Kharonus committed:
[#41340] fixed attachments resource service

- insertCollection now uses upsertMany
- requireCollection no longer returns service instance

oliverguenther reviewed (approved):
The failing spec is unrelated (broken + fixed in dev already). LGTM :+1:
